src/core/tables.js

```diff
@@ -0,0 +1,129 @@
+// @ts-check
+// Module core/tables
+// Handles tables in the document.
+// Generates a List of Tables wherever there is a #list-of-tables element.
+
+import { addId, getIntlData, renameElement, wrapInner } from "./utils.js";
+import { html } from "./import-maps.js";
+
+export const name = "core/tables";
+
+const localizationStrings = {
+  en: {
+    list_of_tables: "List of Tables",
+    table: "Table ",
+  },
+};
+
+const l10n = getIntlData(localizationStrings);
+
+export function run() {
+  const listOfTables = collectTables();
+  const listOfTablesElement = document.querySelector("section#list-of-tables");
+  if (listOfTables.length && listOfTablesElement) {
+    decorateListOfTables(listOfTablesElement);
+    listOfTablesElement.append(
+      html`<h1>${l10n.list_of_tables}</h1>`,
+      html`<ul class="list-of-tables">
+        ${listOfTables}
+      </ul>`
+    );
+  }
+}
+
+/**
+ * process all tables
+ */
+function collectTables() {
+  /** @type {HTMLLIElement[]} */
+  const listOfTables = [];
+  /** @type {NodeListOf<HTMLTableElement>} */
+  const tables = document.querySelectorAll("table.numbered");
+  tables.forEach((table, i) => {
+    const caption = table.querySelector("caption");
+    // there is a separate linter rule to catch numbered tables without captions
+    if (caption) {
+      decorateTable(table, caption, i);
+      listOfTables.push(getListOfTablesListItem(table.id, caption));
+    }
+  });
```

Comment:
Question: If some table doesn't have a caption, we'll jump from Table 1 to Table 3? Should it be table 2 instead?

Reply:
```suggestion
  tables
  // there is a separate linter rule to catch numbered tables without captions
    .filter((table) => !!table.querySelector("caption"))
    .forEach((table, i) => {
      const caption = table.querySelector("caption");
      decorateTable(table, caption, i);
      listOfTables.push(getListOfTablesListItem(table.id, caption));
  });
```

Reply:
I think this solves the same problem that class="numbered" already solved. But perhaps this solves it in a more elegant way.

If someone is using a table for layout (for example), that table should not be included in the list of tables. Having a per-table opt-in of class="numbered" is a way to indicate this is a genuine table with tabular data.

Ideally, all of these good tables have a caption. They can later be referenced by table number or caption. If a table does not have a caption, it can only be referenced by table number. This matters when generating the list of tables.

We decided not to generate an error and only make a linter rule for captionless tables. I interpreted this as being lenient on the caption requirement. But if we decide not to consider captionless tables at all, we don't need to worry about how they are referenced in the list of tables.

Further, someone using tables for non-tabular data is unlikely to provide a caption. So maybe we can skip the class="numbered" requirement and instead replace it with a caption requirement.

What do you think?
(I approve of this change either way. But if we accept it, we should probably make other changes to reflect its impact.)

Reply:
Correction: It was changed from a warning to a linter rule. Not from an error to a linter rule.